controller/instance_manager_controller.go

```diff
@@ -533,74 +568,106 @@ func (imc *InstanceManagerController) cleanUpPDBForNonExistingIM() error {
+		return false, "", fmt.Errorf("the instance manager %v has invalid type: %v ", im.Name, im.Spec.Type)
 	}
 
-	// Must wait for all volumes detached from the current node first.
-	// This also means that we must wait until the PDB of engine instance manager on the current node is deleted
-	allVolumeDetached, err := imc.areAllVolumesDetachedFromCurrentNode()
+	// Must wait for all instance removed from the current node first.
+	allInstanceRemoved, err := imc.areAllInstanceRemovedFromCurrentNode(im)
 	if err != nil {
-		return false, err
+		return false, "", err
 	}
-	if !allVolumeDetached {
-		return false, nil
+	if !allInstanceRemoved {
+		return false, "", nil
 	}
 
 	allowDrainingNodeWithLastReplica, err := imc.ds.GetSettingAsBool(types.SettingNameAllowNodeDrainWithLastHealthyReplica)
 	if err != nil {
-		return false, err
+		return false, "", err
 	}
 	if allowDrainingNodeWithLastReplica {
-		return true, nil
+		return true, "", nil
 	}
 
 	replicasOnCurrentNode, err := imc.ds.ListReplicasByNodeRO(imc.controllerID)
 	if err != nil {
 		if datastore.ErrorIsNotFound(err) {
-			return true, nil
+			return true, "", nil
 		}
-		return false, err
+		return false, "", err
 	}
 
 	// For each replica process in the current node,
 	// find out whether there is a healthy replica of the same volume on another node
+	availableNode := ""
 	for _, replica := range replicasOnCurrentNode {
 		vol, err := imc.ds.GetVolume(replica.Spec.VolumeName)
 		if err != nil {
-			return false, err
+			return false, "", err
 		}
 
 		replicas, err := imc.ds.ListVolumeReplicas(vol.Name)
 		if err != nil {
-			return false, err
+			return false, "", err
 		}
 
-		hasHealthyReplicaOnAnotherNode := false
+		hasPDBOnAnotherNode := false
 		isUnusedReplicaOnCurrentNode := false
 		for _, r := range replicas {
 			if r.Spec.HealthyAt != "" && r.Spec.FailedAt == "" && r.Spec.NodeID != imc.controllerID {
-				hasHealthyReplicaOnAnotherNode = true
-				break
+				unschedulable, err := imc.ds.IsKubeNodeUnschedulable(r.Spec.NodeID)
+				if err != nil {
+					return false, "", err
+				}
+				if unschedulable {
+					continue
+				}
+				availableNode = r.Spec.NodeID
+
+				var rIM *longhorn.InstanceManager
+				if r.Status.InstanceManagerName == "" {
+					rIM, err = imc.ds.GetInstanceManagerByInstance(r)
+					if err != nil && !types.ErrorIsNotFound(err) {
+						return false, "", err
+					}
+				} else {
+					rIM, err = imc.ds.GetInstanceManager(r.Status.InstanceManagerName)
+					if err != nil && !apierrors.IsNotFound(err) {
+						return false, "", err
+					}
+				}
+				if rIM == nil || rIM.Status.CurrentState != longhorn.InstanceManagerStateRunning {
+					continue
+				}
+				pdb, err := imc.ds.GetPDBRO(imc.getPDBName(rIM))
+				if err != nil && !datastore.ErrorIsNotFound(err) {
+					return false, "", err
+				}
+				if pdb != nil {
+					hasPDBOnAnotherNode = true
+					break
+				}
 			}
```

Comment:
1. Can we refactor this block like the following?
```
containOtherHealthyReplicas := r.Spec.HealthyAt != "" && r.Spec.FailedAt == "" && r.Spec.NodeID != imc.controllerID
isOtherHealthyReplicaOnScheduableNode := ....
isOtherHealthyReplicaHasRuningIM := ....
isOtherHealthyReplicaHasIMPDB := 
```
2. Can the logic handle this corner case: A volume has 2 replicas on node2 and node3, and the volume is attached to node1. Then the node2 & node3 are drained (almost) simultaneously.
3. What if the controller does not enqueue `availableNode`? (This output field is a little bit confusing)

Reply:
> 1. Can we refactor this block like the following?
> 
> ```
> containOtherHealthyReplicas := r.Spec.HealthyAt != "" && r.Spec.FailedAt == "" && r.Spec.NodeID != imc.controllerID
> isOtherHealthyReplicaOnScheduableNode := ....
> isOtherHealthyReplicaHasRuningIM := ....
> isOtherHealthyReplicaHasIMPDB := 
> ```

I've made some refactoring. PTAL.

> 
> 2. Can the logic handle this corner case: A volume has 2 replicas on node2 and node3, and the volume is attached to node1. Then the node2 & node3 are drained (almost) simultaneously.

- We do not delete PDB when a node has attached volume.
- We are not expecting CA to remove multiple nodes at once. Ref [How does scale-down work?](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-does-scale-down-work)
```
Cluster Autoscaler terminates one non-empty node at a time to reduce the risk of creating new unschedulable pods.
```
- CA would have made a decision to not scale down at pre-check if it's the last healthy replica available. So this logic applies to manual cordon/drain.

This logic does not handle the case of manual cordon/drain of multiple nodes simultaneously. Any suggestion to handle this? 

> 3. What if the controller does not enqueue `availableNode`? (This output field is a little bit confusing)
We will block PDB deletion.

Reply:
@shuo-wu @innobead , I've removed logic to enqueue another schedulable node. Don't think we need to re-create the PDB.